File: src/MainClasses/Photo.java
package MainClasses;

import java.time.LocalDateTime;
import java.util.*;

public class Photo extends BaseClass<Photo>{

    private String ownerId;

    private String photoName;
    private String title;

    private Set<String> tags;

    private String caption;

    private Boolean isFavorable;

    private LocalDateTime lastModified;

    private Set<String> commentIds = new HashSet<>();

    private Set<String> albumIds = new HashSet<>();
    private Set<String> postIds = new HashSet<>();

    private final LocalDateTime createdAt;

    void updateTime(){
        lastModified = LocalDateTime.now();
    }

    public Photo(String ownerId, String photoName, Set<String> albumIds , Set<String> tags, String caption, Boolean isFavorable, String title) {
        this.ownerId = ownerId;
        this.photoName = photoName;
        this.tags = tags;
        this.caption = caption;
        this.isFavorable = isFavorable;
        this.albumIds = albumIds;
        createdAt = LocalDateTime.now();
        this.title = title;
    }


    public String getOwnerId() {
        return ownerId;
    }

    public void setOwnerId(String ownerId) {
        this.ownerId = ownerId;
        updateTime();
    }

    public String getPhotoName() {
        return photoName;
    }

    public void setPhotoName(String photoName) {
        this.photoName = photoName;
        updateTime();
    }

    public Set<String> getTags() {
        return tags;
    }

    public void setTags(Set<String> tags) {
        this.tags = tags;
        updateTime();
    }

    public String getCaption() {
        return caption;
    }

    public void setCaption(String caption) {
        this.caption = caption;
        updateTime();
    }

    public Boolean getFavorable() {
        return isFavorable;
    }

    public void setFavorable(Boolean favorable) {
        isFavorable = favorable;
    }

    public Set<String> getCommentIds() {
        return commentIds;
    }

    public void setCommentIds(Set<String> commentIds) {
        this.commentIds = commentIds;
        updateTime();
    }

    public Set<String> getAlbumIds() {
        return albumIds;
    }

    public void setAlbumIds(Set<String> photoAlbumIds) {
        this.albumIds = photoAlbumIds;
        updateTime();
    }

    public Set<String> getPostIds() {
        return postIds;
    }
    public void setPostIds(Set<String> postIds) {
        this.postIds = postIds;
    }

    public LocalDateTime getLastModified() {
        return lastModified;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        Photo photo = (Photo) o;
        return Objects.equals(getId(), photo.getId());
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(getId());
    }


}

File: frontend/lib/models/photo.dart
class Photo {
  const Photo({
    required this.id,
    required this.ownerId,
    required this.photoName,
    this.title = '',
    this.tags = const {},
    this.caption = '',
    this.isFavorable = false,
    this.lastModified,
    this.commentIds = const {},
    this.albumIds = const {},
    this.postIds = const {},
    required this.createdAt,
  });

  final String id;
  final String ownerId;
  final String photoName;
  final String title;
  final Set<String> tags;
  final String caption;
  final bool isFavorable;
  final DateTime? lastModified;
  final Set<String> commentIds;
  final Set<String> albumIds;
  final Set<String> postIds;
  final DateTime createdAt;

  Photo copyWith({
    String? id,
    String? ownerId,
    String? photoName,
    String? title,
    Set<String>? tags,
    String? caption,
    bool? isFavorable,
    DateTime? lastModified,
    Set<String>? commentIds,
    Set<String>? albumIds,
    Set<String>? postIds,
    DateTime? createdAt,
  }) {
    return Photo(
      id: id ?? this.id,
      ownerId: ownerId ?? this.ownerId,
      photoName: photoName ?? this.photoName,
      title: title ?? this.title,
      tags: tags ?? this.tags,
      caption: caption ?? this.caption,
      isFavorable: isFavorable ?? this.isFavorable,
      lastModified: lastModified ?? this.lastModified,
      commentIds: commentIds ?? this.commentIds,
      albumIds: albumIds ?? this.albumIds,
      postIds: postIds ?? this.postIds,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  factory Photo.fromJson(Map<String, dynamic> json) {
    return Photo(
      id: json['id'] as String,
      ownerId: json['ownerId'] as String,
      photoName: json['photoName'] as String,
      title: json['title'] as String? ?? '',
      tags: Set<String>.from(json['tags'] ?? const []),
      caption: json['caption'] as String? ?? '',
      isFavorable: (json['isFavorable'] ?? json['favorable']) as bool? ?? false,
      lastModified: DateTime.now(),
      commentIds: Set<String>.from(json['commentIds'] ?? const []),
      albumIds: Set<String>.from(json['albumIds'] ?? const []),
      postIds: Set<String>.from(json['postIds'] ?? const []),
      createdAt: DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'ownerId': ownerId,
      'photoName': photoName,
      'title': title,
      'tags': tags.toList(),
      'caption': caption,
      'isFavorable': isFavorable,
      'lastModified': lastModified?.toIso8601String(),
      'commentIds': commentIds.toList(),
      'albumIds': albumIds.toList(),
      'postIds': postIds.toList(),
      'createdAt': createdAt.toIso8601String(),
    };
  }
}
